core/src/main/scala/cell/Cell.scala

```diff
@@ -611,108 +608,103 @@ private class CellImpl[K <: Key[V], V](pool: HandlerPool, val key: K, lattice: L
 
 }
 
-private class CompleteDepRunnable[K <: Key[V], V](
-  val pool: HandlerPool,
-  val cell: Cell[K, V],
-  val shortCutValueCallback: V => Outcome[V],
-  val completer: CellCompleter[K, V])
-  extends Runnable with OnCompleteRunnable with (Try[V] => Unit) {
-  // must be filled in before running it
-  var value: Try[V] = null
-
-  override def apply(x: Try[V]): Unit = x match {
-    case Success(v) =>
-      shortCutValueCallback(v) match {
-        case FinalOutcome(v) => completer.putFinal(v)
-        case NextOutcome(v) => completer.putNext(v)
-        case NoOutcome =>
-          completer.removeDep(cell)
-          completer.removeNextDep(cell)
-      }
-    case Failure(e) =>
-      completer.removeDep(cell)
-      completer.removeNextDep(cell)
-  }
-
-  override def run(): Unit = {
-    try apply(value) catch { case NonFatal(e) => pool reportFailure e }
-  }
-
-  def executeWithValue(v: Try[V]): Unit = {
-    value = v
-    try pool.execute(this) catch { case NonFatal(t) => pool reportFailure t }
-  }
-}
-
 // copied from `impl.CallbackRunnable` in Scala core lib.
 /**
  * @param executor   The thread that runs the callback function
  * @param onComplete Callback function that is triggered on an onComplete event
- * @param cell       The cell that depends on this callback
+ * @param dependentCell       The cell that depends on this callback
  */
-private class CompleteCallbackRunnable[K <: Key[V], V](val executor: HandlerPool, val onComplete: Try[V] => Any, val cell: Cell[K, V]) extends Runnable with OnCompleteRunnable {
+private class CompleteCallbackRunnable[K <: Key[V], V](
+  val executor: HandlerPool,
+  val dependentCell: Cell[K, V],
+  val onComplete: Try[V] => Any) {
   // must be filled in before running it
-  var value: Try[V] = null
-
-  override def run() = {
-    require(value ne null) // must set value to non-null before running!
-    try onComplete(value) catch { case NonFatal(e) => executor reportFailure e }
-  }
+  var started: Boolean = false
 
   def executeWithValue(v: Try[V]): Unit = {
-    require(value eq null) // can't complete it twice
-    value = v
+    require(!started) // can't complete it twice
+    started = true
     // Note that we cannot prepare the ExecutionContext at this point, since we might
     // already be running on a different thread!
-    try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
+    try executor.execute(() => {
+      onComplete(v); ()
+    }) catch {
```

Comment:
I'm not sure it's a good idea to create a new closure here; I would expect it to affect performance negatively. Why not do it as before and (a) let `CompleteCallbackRunnable` extend `Runnable` and (b) submit `this`?

Reply:
This is how it had already been implemented in NextCallbackRunnable. I'd like both classes to be alike as they do very similar things.
3 of the classes implement Runnable but only 1 class uses run() (as stated by a coverage analysis) so I decided to remove the Runnable interface in favor of how NextCallbackRunnable does it.

Reply:
It looks like `NextCallbackRunnable` was not really implemented in the optimal way then. Since you are reworking these classes anyway, I suggest to do it right this time, and avoid closure allocations that are not necessary. In general, Reactive Async should be as close to the performance of regular futures as possible. However, we won't be able to achieve this if we give up on possible performance improvements. I don't want to merge code where we already know that it can be done better.